### ISSUE TITLE
FBA Inbound API migration

### ISSUE THREAD

**Issue by NoxArt** [jlevers/selling-partner-api#823]
Hi, I know this is mainly for reporting issues and we're not entitled to ask for this, but does anyone have any guide / sample / advice about converting from old jlevers library & FBA Inbound v0 to new jlevers library & FBA Inbound v2024-03-20? I kinda understand the new library and already migrated other APIs, but the FBA Inbound changes I can't wrap my head around and time is running short

Thank you for any help!

**Comment by dartheditous** [comment]
I'm working on this as well, although I'm not using jlevers' library. If you have a specific question I'll let you know if I've got that far.

It really does seem to be unnecessarily complicated. For example, there are skus which Amazon already knows need no prep. For these, you can specify prepOwner as "NONE". But then there are skus which Amazon doesn't know the prep details for (even though they never needed any under the v0 API). You can update these yourself with setPrepDetails, by setting prepCategory to "NONE", but you also have to add a prepType of "ITEM_NO_PREP", and when you create the shipment, you have to specify the prepOwner as "SELLER", even though no prep is required! 🤦‍♂️

**Comment by jlevers** [comment]
I just finished implementing a full small parcel inbounding flow, so if you have any particular questions I'm happy to answer them -- I don't have any guide written or anything like that, unfortunately!

**Comment by NoxArt** [comment]
That's great ... would it be possible to post an (anonymized) example? The readme.md already has examples for Uploading a feed document and Downloading a feed result document, so maybe add it among these? I think it would be helpful to many people

**Comment by NoxArt** [comment]
that's understandable, thank you, we will consider it

**Comment by NoxArt** [comment]
Around 50% of the time generatePackingOptions fails with "ERROR: Operation GeneratePackingOptions cannot be processed on an empty inbound plan. Please check again later if the inbound plan", are you experiencing that too?

A thing I can't figure out atm is if they already implemented setting Prep/owner label via API so that it no longer says that I can't use Seller Prep option "does not require labelOwner but SELLER was assigned. Accepted values: [NONE]"

They have this
https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-reference#setprepdetails

But it doesn't seem to help (though I set both category and type to None, not sure if it's correct)

Also I'm getting empty result for listPackingOptions and calling setprepdetails didnt help ... or is that expected?

**Comment by dartheditous** [comment]
> A thing I can't figure out atm is if they already implemented setting Prep/owner label via API so that it no longer says that I can't use Seller Prep option "does not require labelOwner but SELLER was assigned. Accepted values: [NONE]"

Can't quite work out what you're saying here... but you can call listPrepDetails to work out whether you need to set labelOwner and prepOwner to NONE or SELLER (it's really illogical and complicated though. It will give you a list of prepTypes, and if one of them is labelling then you need to set labelOwner to SELLER. If any _other_ prepTypes are present, then you have to set prepOwner to SELLER - even if the only other prepType is ITEM_NO_PREP. And that's without considering any of the constraints, which don't seem to be explained in the docs anywhere).

> But it doesn't seem to help (though I set both category and type to None, not sure if it's correct)

NONE isn't valid for prepType. Did you get the OperationId from your call to setPrepDetails and confirm that the operation succeeded (using getInboundOperationStatus)?

**Comment by NoxArt** [comment]
Thank you for quick reply

Sorry, yes, I mean NONE for Category and [ ITEM_NO_PREP ] for Type

It now passes with labelOwner = NONE, prepOwner = SELLER ... not sure this is what is correct for us, I would have assumed there should be some labelOwner and it would probably be us (seller)

I added checking for operation status + waiting and now it's working, thank you so much

**Comment by NoxArt** [comment]
Hmm, some SKUs seem to require labelOwner = NONE, some require SELLER and createInboundPlan keeps failing on that I don't know how to determine which ones require which or how to change it so it's the same for all ... there is setPrepDetails for prep, but I see nothing for labels

And the terminology is really unclear to me. What exactly is "Packing information" in the diagram on page https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-use-case-guide#fulfillment-inbound-shipment-workflow ?

If I go the YES route - I select a packing option which gives me packing groups containing items ... the packing group not a box? Or does Amazon need the box info if it already told me so? Or 1 group can consist of multiple boxes? And we need to make sure the contents of the box doesn't belong to multiple groups?
What's the difference between a packing group and a shipment?

Their docs are pretty comprehensive and yet really insufficient for me...

**Comment by dartheditous** [comment]
Use listPrepDetails - if ITEM_LABELING is **not** among the prep details for a sku, then (and only then) you can use labelOwner = NONE in createInboundPlan.

It really is a very silly system...

I don't think you can change whether item labelling is required for a sku via the API. You only get the choice when you create the listing, and only on a minority of ASINs, in my experience.

**Comment by NoxArt** [comment]
Thank you!

**Comment by NoxArt** [comment]
Hi,
* in V0 we were using NonPartneredSmallParcelDataInput type=SP carrier=Other
* in v2024 I createInboundPlan with EU marketplace, then generatePlacementOptions with empty request
* then the list of placementOptions is only 1 option
* we confirm it
* then in generateTransportOptions -> listOfTransportationOptions there are all only FREIGHT_LTL and no GROUND_SMALL_PARCEL + also 99% of offered carriers are from China ... this seems wrong? I have no idea what I'm doing wrong ... would you have any idea?

Thank you

**Comment by NoxArt** [comment]
@dartheditous yes I've managed to implement it almost fully, just about to test the very last step, thank you
(it was a big pain though)
